File: 008_StringToInteger.py

#74.06%
class Solution(object):
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        rst = 0
        #Remove left space characters
        str = str.lstrip()
        if not len(str):
            return 0
        #Check first character
        #True as 'positive', False as 'Negative'
        flag = (-1) if str[0] == '-' else 1
        if str[0] == '+' or str[0] == '-':
            startIndex = 1
        else:
            startIndex = 0
        
        INT_MAX = 0x7fffffff
        INT_MIN = 0 - 0x80000000
        #Main routine
        for eachChar in str[startIndex:]:
            if eachChar.isdigit():
                rst = rst * 10 + int(eachChar)
                rstVal = rst * flag
                if rstVal >= INT_MAX:
                    return INT_MAX
                elif rstVal <= INT_MIN:
                    return INT_MIN
            else:
                break
        
        return rst * flag
        
        
        
        

File: 008_StringToInteger.c
//10.87
int myAtoi(char* str) {
    //Check boundary conditions
	if(!str || !strlen(str))
		return 0;
	
	//Find first un-blank character
	char *ptr = str;
	while(*ptr == ' ')
		ptr++;
	long rst = 0;
	bool flag = true; // false while encounter '-', true otherwise
	
	//Check first non-blank character
	char firstChar = *ptr;
	if(firstChar == '-')
	{
		flag = false;
		ptr++;
	}
	else if(firstChar == '+')
	{
		ptr++;
	}
	else if(firstChar < '0' || firstChar > '9')
		return 0;
	
	
	//Check the second integer unit
	//char secChar = *ptr;
	//if(secChar < '0' || secChar > '9')
		//return 0;
	
	//Process the string iteratively
	while(*ptr != '\0')
	{
		char currChar = *ptr;
		if(currChar >= '0' && currChar <= '9')
		{
			int currUnit = currChar - '0';	
			rst *= 10;
			rst += currUnit;
			ptr++;
			//Check if it is overflow
			if(rst < INT_MAX)
				continue;
			else if(flag)
				return INT_MAX;
			else if(-rst <= INT_MIN)
				return INT_MIN;
		}
		else
			break;
	}
	
	if(!flag)
		rst = -rst;
	return rst;		
}

int myAtoi001(char* str) {
    //Check boundary conditions
	if(!str || !strlen(str))
		return 0;
	
	//Find first un-blank character
	char *ptr = str;
	while(*ptr == ' ')
		ptr++;
	long rst = 0;
	bool flag = true; // false while encounter '-', true otherwise
	
	//Check first non-blank character
	char firstChar = *ptr;
	if(firstChar == '-')
	{
		flag = false;
		ptr++;
	}
	else if(firstChar == '+')
	{
		ptr++;
	}
	else if(firstChar < '0' || firstChar > '9')
		return 0;
	
	
	//Check the second integer unit
	char secChar = *ptr;
	if(secChar < '0' || secChar > '9')
		return 0;
	
	//Process the string iteratively
	while(*ptr != '\0')
	{
		char currChar = *ptr;
		if(currChar >= '0' && currChar <= '9')
		{
			int currUnit = currChar - '0';	
			rst *= 10;
			rst += currUnit;
			//Check if it is overflow
			if(flag && rst >= INT_MAX)
				return INT_MAX;
			if(!flag && -rst <= INT_MIN)
				return INT_MIN;
			ptr++;
		}
		else
		{
			//ptr++;
			break;
		}
	}
	
	if(!flag)
		rst = -rst;
	return rst;		
}